File: python/tests/test_codec.py
"""Codec byte-exactness and round-trip tests (schemas/FORMAT.md)."""

import hashlib
import struct
import zlib

import pytest

from conftest import mkev
from iap.core import codec
from iap.core.events import EventType, MarketEvent, Side


def _sample_events(n=20):
    evs = []
    for i in range(1, n + 1):
        evs.append(
            mkev(i, EventType.ADD, i % 2, 2400 + i, 100 * i, order_id=1000 + i)
        )
    return evs


EV = MarketEvent(
    event_id=1,
    instrument_id=2,
    venue_id=3,
    exchange_ts=4,
    receive_ts=5,
    sequence=6,
    event_type=7,
    side=1,
    price_ticks=-8,
    qty=9,
    order_id=10,
    trade_id=11,
)


def test_jsonl_line_exact_bytes():
    assert codec.encode_jsonl_line(EV) == (
        '{"event_id":1,"instrument_id":2,"venue_id":3,"exchange_ts":4,'
        '"receive_ts":5,"sequence":6,"event_type":7,"side":1,'
        '"price_ticks":-8,"qty":9,"order_id":10,"trade_id":11}'
    )


def test_jsonl_round_trip():
    for ev in _sample_events():
        assert codec.decode_jsonl_line(codec.encode_jsonl_line(ev)) == ev


def test_jsonl_rejects_missing_key():
    line = '{"event_id":1}'
    with pytest.raises(ValueError, match="missing"):
        codec.decode_jsonl_line(line)


def _reject_cases(golden_dir):
    """Parse tests/golden/jsonl_reject_cases.txt -> (reject lines, accept lines)."""
    reject, accept = [], []
    target = reject
    for raw in (golden_dir / "jsonl_reject_cases.txt").read_text().splitlines():
        if raw.strip() == "# ACCEPT":
            target = accept
            continue
        if not raw.strip() or raw.startswith("#"):
            continue
        target.append(raw)
    return reject, accept


def test_jsonl_strictness_parity_fixture(golden_dir):
    """Python rejects exactly the shared fixture lines every port rejects."""
    reject, accept = _reject_cases(golden_dir)
    assert len(reject) >= 25 and len(accept) >= 5
    for line in reject:
        with pytest.raises(ValueError):
            codec.decode_jsonl_line(line)
    for line in accept:
        ev = codec.decode_jsonl_line(line)
        assert codec.decode_jsonl_line(codec.encode_jsonl_line(ev)) == ev


def test_jsonl_rejects_out_of_range_and_duplicate_keys():
    base = codec.encode_jsonl_line(EV)
    with pytest.raises(ValueError, match="event_id"):
        codec.decode_jsonl_line(base.replace('"event_id":1', '"event_id":18446744073709551616'))
    with pytest.raises(ValueError, match="venue_id"):
        codec.decode_jsonl_line(base.replace('"venue_id":3', '"venue_id":70000'))
    with pytest.raises(ValueError, match="side"):
        codec.decode_jsonl_line(base.replace('"side":1', '"side":-1'))
    with pytest.raises(ValueError):
        codec.decode_jsonl_line(base.replace('"trade_id":11', '"trade_id":0,"trade_id":77'))
    with pytest.raises(ValueError, match="non-negative"):
        codec.decode_jsonl_line(base.replace('"sequence":6', '"sequence":-0'))
    assert codec.decode_jsonl_line(base.replace('"price_ticks":-8', '"price_ticks":-0')).price_ticks == 0


def test_iap1_empty_vector_round_trip():
    assert codec.decode_iap1(codec.encode_iap1([])) == []


def test_jsonl_rejects_extra_key():
    line = codec.encode_jsonl_line(EV)[:-1] + ',"extra":1}'
    with pytest.raises(ValueError, match="extra"):
        codec.decode_jsonl_line(line)


def test_jsonl_rejects_wrong_key_order():
    d = EV.to_dict()
    keys = list(d)
    keys[0], keys[1] = keys[1], keys[0]
    import json

    line = json.dumps({k: d[k] for k in keys}, separators=(",", ":"))
    with pytest.raises(ValueError):
        codec.decode_jsonl_line(line)


def test_jsonl_rejects_non_integer_values():
    line = codec.encode_jsonl_line(EV).replace('"qty":9', '"qty":9.0')
    with pytest.raises(ValueError, match="integer"):
        codec.decode_jsonl_line(line)
    line = codec.encode_jsonl_line(EV).replace('"qty":9', '"qty":true')
    with pytest.raises(ValueError, match="integer"):
        codec.decode_jsonl_line(line)


def test_jsonl_rejects_malformed_line():
    with pytest.raises(ValueError):
        codec.decode_jsonl_line("not json at all {")


def test_iap1_sizes_pinned():
    assert codec.IAP1_HEADER_SIZE == 16
    assert codec.IAP1_RECORD_SIZE == 72
    assert codec.IAP1_TRAILER_SIZE == 16
    data = codec.encode_iap1(_sample_events(5))
    assert len(data) == 16 + 5 * 72 + 16


def test_iap1_header_and_trailer_bytes_exact():
    """Empty vector: header + trailer only; CRC-32 of the 16 header bytes."""
    data = codec.encode_iap1([])
    header = struct.pack("<IIQ", 0x49415031, 2, 0)
    assert data == header + struct.pack("<IIQ", zlib.crc32(header), 0, 0)
    assert data[:4] == b"1PAI"  # little-endian magic spells IAP1 reversed
    assert codec.decode_iap1(data) == []


def test_iap1_record_bytes_exact():
    data = codec.encode_iap1([EV])
    body = struct.pack("<IIQ", 0x49415031, 2, 1) + struct.pack(
        "<QIHBBqqQqqQQ", 1, 2, 3, 7, 1, 4, 5, 6, -8, 9, 10, 11
    )
    assert data == body + struct.pack("<IIQ", zlib.crc32(body), 0, 1)


def test_iap1_crc32_known_answer():
    """CRC-32 (IEEE 802.3 / zlib) pinned by a known answer every port must match."""
    assert codec.crc32(b"123456789") == 0xCBF43926
    assert codec.crc32(b"") == 0


def test_iap1_legacy_v1_accepted_without_integrity():
    body = struct.pack("<IIQ", 0x49415031, 1, 1) + struct.pack(
        "<QIHBBqqQqqQQ", 1, 2, 3, 7, 1, 4, 5, 6, -8, 9, 10, 11
    )
    decoded = codec.decode_iap1_ex(body)
    assert decoded.events == [EV]
    assert decoded.version == 1 and decoded.integrity_checked is False
    v2 = codec.decode_iap1_ex(codec.encode_iap1([EV]))
    assert v2.version == 2 and v2.integrity_checked is True


def test_iap1_rejects_corrupted_record_via_crc():
    """A single flipped byte inside a record is detected (fail closed)."""
    data = bytearray(codec.encode_iap1(_sample_events(20)))
    data[16 + 72 * 7 + 50] ^= 0x01  # qty byte of record 7
    with pytest.raises(ValueError, match="CRC-32"):
        codec.decode_iap1(bytes(data))
    data = bytearray(codec.encode_iap1(_sample_events(3)))
    data[-8] ^= 0x01  # count echo
    with pytest.raises(ValueError, match="count echo"):
        codec.decode_iap1(bytes(data))
    data = bytearray(codec.encode_iap1(_sample_events(3)))
    data[-12] = 1  # reserved field
    with pytest.raises(ValueError, match="reserved"):
        codec.decode_iap1(bytes(data))


def test_encode_iap1_rejects_out_of_domain_with_value_error():
    bad = MarketEvent(**{**EV.to_dict(), "venue_id": 70000})
    with pytest.raises(ValueError, match="venue_id"):
        codec.encode_iap1([EV, bad])
    bad = MarketEvent(**{**EV.to_dict(), "event_id": 1 << 64})
    with pytest.raises(ValueError, match="event 0"):
        codec.encode_iap1([bad])


def test_iap1_round_trip():
    evs = _sample_events(50) + [EV]
    assert codec.decode_iap1(codec.encode_iap1(evs)) == evs


def test_iap1_rejects_bad_magic():
    data = bytearray(codec.encode_iap1([EV]))
    data[0] ^= 0xFF
    with pytest.raises(ValueError, match="magic"):
        codec.decode_iap1(bytes(data))


def test_iap1_rejects_bad_version():
    data = struct.pack("<IIQ", 0x49415031, 3, 0)
    with pytest.raises(ValueError, match="version"):
        codec.decode_iap1(data)


def test_iap1_rejects_truncation_and_count_mismatch():
    data = codec.encode_iap1(_sample_events(3))
    with pytest.raises(ValueError):
        codec.decode_iap1(data[:-1])
    with pytest.raises(ValueError):
        codec.decode_iap1(data + b"\x00" * 72)
    with pytest.raises(ValueError):
        codec.decode_iap1(data[:10])


def test_file_round_trips(tmp_path):
    evs = _sample_events(30)
    jl = tmp_path / "x.jsonl"
    bi = tmp_path / "x.iap1"
    assert codec.write_jsonl(jl, evs) == 30
    assert codec.write_iap1(bi, evs) == 30
    assert codec.read_jsonl(jl) == evs
    assert codec.read_iap1(bi) == evs


def test_jsonl_and_iap1_carry_identical_events():
    evs = _sample_events(10)
    via_json = [codec.decode_jsonl_line(codec.encode_jsonl_line(e)) for e in evs]
    via_bin = codec.decode_iap1(codec.encode_iap1(evs))
    assert via_json == via_bin == evs


def test_sha256_helpers(tmp_path):
    evs = _sample_events(5)
    data = codec.encode_iap1(evs)
    assert codec.sha256_bytes(data) == hashlib.sha256(data).hexdigest()
    assert codec.sha256_events_iap1(evs) == hashlib.sha256(data).hexdigest()
    p = tmp_path / "f.iap1"
    p.write_bytes(data)
    assert codec.sha256_file(p) == hashlib.sha256(data).hexdigest()


def test_encode_jsonl_bytes_lf_terminated():
    data = codec.encode_jsonl(_sample_events(3))
    assert data.endswith(b"\n") and data.count(b"\n") == 3

File: cpp/tests/test_codec.cpp
// JSONL + IAP1 codec unit tests (schemas/FORMAT.md; API_CORE.md section 3).

#include <gtest/gtest.h>

#include <cstdint>
#include <stdexcept>
#include <string>
#include <vector>

#include "iap/marketdata/codec.hpp"
#include "iap/marketdata/events.hpp"
#include "sha256.hpp"

using iap::decode_iap1;
using iap::decode_jsonl;
using iap::decode_jsonl_line;
using iap::encode_iap1;
using iap::encode_jsonl;
using iap::encode_jsonl_line;
using iap::EventType;
using iap::MarketEvent;

namespace {

MarketEvent sample() {
    return MarketEvent::of(7, 1, 1, 1787578200000000000LL,
                           1787578200000181199LL, 9,
                           static_cast<std::uint8_t>(EventType::ADD), 0, 2450,
                           300, 1234567890123456789ULL, 0);
}

std::vector<MarketEvent> sample_vec() {
    std::vector<MarketEvent> v;
    v.push_back(sample());
    MarketEvent e2 = sample();
    e2.event_id = 8;
    e2.sequence = 10;
    e2.side = 1;
    e2.price_ticks = 2451;
    v.push_back(e2);
    return v;
}

}  // namespace

// ------------------------------------------------------------------- JSONL

TEST(Jsonl, EncodeLineCanonicalForm) {
    std::string line = encode_jsonl_line(sample());
    EXPECT_EQ(line,
              "{\"event_id\":7,\"instrument_id\":1,\"venue_id\":1,"
              "\"exchange_ts\":1787578200000000000,"
              "\"receive_ts\":1787578200000181199,\"sequence\":9,"
              "\"event_type\":1,\"side\":0,\"price_ticks\":2450,\"qty\":300,"
              "\"order_id\":1234567890123456789,\"trade_id\":0}");
}

TEST(Jsonl, RoundTripLine) {
    MarketEvent ev = sample();
    EXPECT_EQ(decode_jsonl_line(encode_jsonl_line(ev)), ev);
}

TEST(Jsonl, RoundTripNegativeAndExtremeValues) {
    MarketEvent ev = sample();
    ev.exchange_ts = -42;
    ev.receive_ts = -1;
    ev.price_ticks = INT64_MIN;
    ev.qty = INT64_MAX;
    ev.order_id = UINT64_MAX;
    EXPECT_EQ(decode_jsonl_line(encode_jsonl_line(ev)), ev);
}

TEST(Jsonl, EncodeVectorHasTrailingLf) {
    std::string data = encode_jsonl(sample_vec());
    ASSERT_FALSE(data.empty());
    EXPECT_EQ(data.back(), '\n');
    EXPECT_EQ(decode_jsonl(data).size(), 2u);
}

TEST(Jsonl, DecodeSkipsBlankLines) {
    std::string data = encode_jsonl_line(sample()) + "\n\n  \n" +
                       encode_jsonl_line(sample()) + "\n";
    EXPECT_EQ(decode_jsonl(data).size(), 2u);
}

TEST(Jsonl, RejectsMisorderedKeys) {
    // instrument_id and event_id swapped.
    std::string line =
        "{\"instrument_id\":1,\"event_id\":7,\"venue_id\":1,"
        "\"exchange_ts\":0,\"receive_ts\":0,\"sequence\":9,\"event_type\":1,"
        "\"side\":0,\"price_ticks\":2450,\"qty\":300,\"order_id\":5,"
        "\"trade_id\":0}";
    EXPECT_THROW(decode_jsonl_line(line), std::invalid_argument);
}

TEST(Jsonl, RejectsMissingKey) {
    std::string line = encode_jsonl_line(sample());
    auto pos = line.find(",\"trade_id\":0");
    line.erase(pos, std::string(",\"trade_id\":0").size());
    EXPECT_THROW(decode_jsonl_line(line), std::invalid_argument);
}

TEST(Jsonl, RejectsExtraKey) {
    std::string line = encode_jsonl_line(sample());
    line.insert(line.size() - 1, ",\"extra\":1");
    EXPECT_THROW(decode_jsonl_line(line), std::invalid_argument);
}

TEST(Jsonl, RejectsNonIntegerValues) {
    std::string base = encode_jsonl_line(sample());
    {
        std::string line = base;
        line.replace(line.find("\"qty\":300"), 9, "\"qty\":3.5");
        EXPECT_THROW(decode_jsonl_line(line), std::invalid_argument);
    }
    {
        std::string line = base;
        line.replace(line.find("\"qty\":300"), 9, "\"qty\":true");
        EXPECT_THROW(decode_jsonl_line(line), std::invalid_argument);
    }
    {
        std::string line = base;
        line.replace(line.find("\"qty\":300"), 9, "\"qty\":\"3\"");
        EXPECT_THROW(decode_jsonl_line(line), std::invalid_argument);
    }
    {
        std::string line = base;
        line.replace(line.find("\"qty\":300"), 9, "\"qty\":3e2");
        EXPECT_THROW(decode_jsonl_line(line), std::invalid_argument);
    }
}

TEST(Jsonl, RejectsMalformedLines) {
    EXPECT_THROW(decode_jsonl_line(""), std::invalid_argument);
    EXPECT_THROW(decode_jsonl_line("not json"), std::invalid_argument);
    EXPECT_THROW(decode_jsonl_line("{"), std::invalid_argument);
    std::string line = encode_jsonl_line(sample());
    EXPECT_THROW(decode_jsonl_line(line + "x"), std::invalid_argument);
}

TEST(Jsonl, RejectsUnsignedFieldOutOfRange) {
    std::string line = encode_jsonl_line(sample());
    line.replace(line.find("\"venue_id\":1"), 12, "\"venue_id\":70000");
    EXPECT_THROW(decode_jsonl_line(line), std::invalid_argument);
    std::string line2 = encode_jsonl_line(sample());
    line2.replace(line2.find("\"sequence\":9"), 12, "\"sequence\":-1");
    EXPECT_THROW(decode_jsonl_line(line2), std::invalid_argument);
}

// -------------------------------------------------------------------- IAP1

TEST(Iap1, HeaderAndRecordBytes) {
    auto data = encode_iap1(sample_vec());
    ASSERT_EQ(data.size(), 16u + 2u * 72u);
    // Magic 0x49415031 little-endian on disk: '1' 'P' 'A' 'I'.
    EXPECT_EQ(data[0], '1');
    EXPECT_EQ(data[1], 'P');
    EXPECT_EQ(data[2], 'A');
    EXPECT_EQ(data[3], 'I');
    EXPECT_EQ(data[4], 1u);  // version LE
    EXPECT_EQ(data[5], 0u);
    EXPECT_EQ(data[8], 2u);  // count LE
    // First record starts with event_id=7 LE.
    EXPECT_EQ(data[16], 7u);
    EXPECT_EQ(data[17], 0u);
}

TEST(Iap1, RoundTrip) {
    auto vec = sample_vec();
    auto decoded = decode_iap1(encode_iap1(vec));
    ASSERT_EQ(decoded.size(), vec.size());
    for (std::size_t i = 0; i < vec.size(); ++i) {
        EXPECT_EQ(decoded[i], vec[i]);
    }
}

TEST(Iap1, EmptyVectorRoundTrip) {
    auto data = encode_iap1({});
    EXPECT_EQ(data.size(), 16u);
    EXPECT_TRUE(decode_iap1(data).empty());
}

TEST(Iap1, RejectsBadMagic) {
    auto data = encode_iap1(sample_vec());
    data[0] = 'X';
    EXPECT_THROW(decode_iap1(data), std::invalid_argument);
}

TEST(Iap1, RejectsBadVersion) {
    auto data = encode_iap1(sample_vec());
    data[4] = 2;
    EXPECT_THROW(decode_iap1(data), std::invalid_argument);
}

TEST(Iap1, RejectsTruncatedHeader) {
    std::vector<std::uint8_t> data(10, 0);
    EXPECT_THROW(decode_iap1(data), std::invalid_argument);
}

TEST(Iap1, RejectsSizeCountMismatch) {
    auto data = encode_iap1(sample_vec());
    data.pop_back();  // truncate a record
    EXPECT_THROW(decode_iap1(data), std::invalid_argument);
    auto data2 = encode_iap1(sample_vec());
    data2[8] = 3;  // count says 3, payload has 2
    EXPECT_THROW(decode_iap1(data2), std::invalid_argument);
}

TEST(Iap1, FileRoundTrip) {
    std::string path = ::testing::TempDir() + "iap_codec_roundtrip.iap1";
    auto vec = sample_vec();
    EXPECT_EQ(iap::write_iap1(path, vec), vec.size());
    auto back = iap::read_iap1(path);
    ASSERT_EQ(back.size(), vec.size());
    EXPECT_EQ(back[0], vec[0]);
    EXPECT_EQ(back[1], vec[1]);
}

TEST(Iap1, ReadMissingFileThrows) {
    EXPECT_THROW(iap::read_iap1("/nonexistent/path/x.iap1"),
                 std::runtime_error);
}

// ----------------------------------------------------------------- SHA-256

TEST(Sha256Impl, KnownAnswers) {
    EXPECT_EQ(iap_test::Sha256::hash(std::string("")),
              "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855");
    EXPECT_EQ(iap_test::Sha256::hash(std::string("abc")),
              "ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad");
    // Multi-block message (> 64 bytes).
    EXPECT_EQ(iap_test::Sha256::hash(std::string(
                  "abcdefghbcdefghicdefghijdefghijkefghijklfghijklmghijklmn"
                  "hijklmnoijklmnopjklmnopqklmnopqrlmnopqrsmnopqrstnopqrstu")),
              "cf5b16a778af8380036ce59e7b0492370b249b11e8f07a51afac45037afee9d1");
}
